File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        hasmap ={}
        n = len(strs)
        for i in range(n):
            sorted_str = ''.join(sorted(strs[i]))
            if sorted_str in hasmap:
                hasmap[sorted_str].append(strs[i])
            else:
                hasmap[sorted_str] = [strs[i]]
        return list(hasmap.values())

File: 0049-group-anagrams/0049-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {

    let map = {}

    for (let i = 0; i< strs.length ;i++){
        let sortStr = strs[i].split("").sort().join("")
        if(!map[sortStr]){
            map[sortStr] = [strs[i]]
        }else{
            map[sortStr].push(strs[i])
        }
    }
    return Object.values(map)
    
};



// strs = ["eat","tea","tan","ate","nat","bat"]

// Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

// str => sorted => [aet,aet,ant,aet,ant,abt]


// map {
//     aet : [eat,tea,ate]
//     ant :[tan,ant]
//     abt:[bat]
// }
// return map(values)

// S = nlogn  * m
// S = n * m